contracts/sovereignChains/AgglayerBridgeL2FromEtrog.sol

```diff
@@ -0,0 +1,137 @@
+     */
+    function initializeFromEtrog(
+        address _bridgeManager,
+        address _emergencyBridgePauser,
+        address _emergencyBridgeUnpauser,
+        address _proxiedTokensManager,
+        address[] memory wrappedTokensAddresses,
+        uint128 initNativeSupply
+    ) public virtual getInitializedVersion reinitializer(3) {
+        // Checks that upgrade is being done from the contract initialized
+        if (_initializerVersion == 0) {
+            revert InvalidInitializeFunction();
+        }
+
+        // Checks that the new parameters are not already set
+        if (
+            bridgeManager != address(0) ||
+            emergencyBridgePauser != address(0) ||
+            emergencyBridgeUnpauser != address(0) ||
+            proxiedTokensManager != address(0)
+        ) {
+            revert ParameterAlreadyInitialized();
+        }
+
+        // Set bridge manager
+        bridgeManager = _bridgeManager;
+
+        // Set emergency bridge pauser and unpauser
+        emergencyBridgePauser = _emergencyBridgePauser;
+        emit AcceptEmergencyBridgePauserRole(address(0), emergencyBridgePauser);
+        emergencyBridgeUnpauser = _emergencyBridgeUnpauser;
+        emit AcceptEmergencyBridgeUnpauserRole(
+            address(0),
+            emergencyBridgeUnpauser
+        );
+
+        // Set proxied tokens manager
+        require(
+            _proxiedTokensManager != address(this),
+            BridgeAddressNotAllowed()
+        );
+
+        // It's not allowed proxiedTokensManager to be zero address. If disabling token upgradability is required, add a not owned account like 0xffff...fffff
+        require(_proxiedTokensManager != address(0), InvalidZeroAddress());
+        proxiedTokensManager = _proxiedTokensManager;
+        emit AcceptProxiedTokensManagerRole(address(0), proxiedTokensManager);
+
+        _setLBTFromTokensAddress(wrappedTokensAddresses, initNativeSupply);
+    }
+
+    /**
+     * @dev Builds origin network/address/amount arrays from token addresses (index 0 = native token) and sets the Local Balance Tree.
+     * @param wrappedTokensAddresses Array of wrapped-token contract addresses to include; internal arrays reserve index 0 for the native token, so wrapped-token data is mapped starting at index 1.
+     * @param initNativeSupply Initial supply used to compute the native token amount when WETHToken is unset
+     */
+    function _setLBTFromTokensAddress(
+        address[] memory wrappedTokensAddresses,
+        uint256 initNativeSupply
+    ) internal {
+        uint256 amountWETH;
+        if (address(WETHToken) == address(0)) {
+            uint256 balance = address(this).balance;
+            require(
+                initNativeSupply >= balance,
+                "initNativeSupply < ETH balance"
+            );
+            amountWETH = initNativeSupply - balance;
+        } else {
+            amountWETH = ITokenWrappedBridgeUpgradeable(address(WETHToken))
```

Comment:
i think we need more comments here:
- what happens if we have gast token ( checking gastTokenAddress or WETH token, both are gine9
- what happens if not